audisp-tacplus/regex_helper.c

```diff
@@ -0,0 +1,88 @@
+#include <ctype.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+#include <syslog.h>
+#include <stdlib.h>
+
+#include "user_secret.h"
+#include "trace.h"
+
+#define min(a,b)            (((a) < (b)) ? (a) : (b))
+
+/* 
+ * Macros for user secret regex
+ * These are BRE regex, please refer to: https://en.wikibooks.org/wiki/Regular_Expressions/POSIX_Basic_Regular_Expressions
+ */
+#define USER_SECRET_REGEX_WHITE_SPACE              "[[:space:]]*"
+#define USER_SECRET_REGEX_SECRET                   "\\([^[:space:]]*\\)"
+
+/* Regex match group count, 2 because only have 1 subexpression for user secret */
+#define REGEX_MATCH_GROUP_COUNT      2
+
+/* The user secret mask */
+#define USER_SECRET_MASK                   '*'
+
+/* Replace user secret with regex */
+int remove_user_secret_by_regex(const char* command, char* result_buffer, size_t buffer_size, regex_t regex)
+{
+    regmatch_t pmatch[REGEX_MATCH_GROUP_COUNT];
+    if (regexec(&regex, command, REGEX_MATCH_GROUP_COUNT, pmatch, 0) == REG_NOMATCH) {
+        trace("User command not match.\n");
+        return USER_SECRET_NOT_FOUND;
+    }
+    
+    if (pmatch[1].rm_so < 0) {
+        trace("User secret not found.\n");
+        return USER_SECRET_NOT_FOUND;
+    }
+    
+    /* Found user secret between pmatch[1].rm_so to pmatch[1].rm_eo, replace it. */
+    trace("Found user secret between: %d -- %d\n", pmatch[1].rm_so, pmatch[1].rm_eo);
+    
+    /* Copy user command first. */
+    snprintf(result_buffer, buffer_size, "%s", command);
+
+    /* Replace user secret. */
```

Comment:
>Replace user secret [](http://example.com/codeflow?start=7&length=19)

It is replace but the secret length is leaked. Could you check the requirement on the secret removing?